src/emu/sound.h

```diff
@@ -6,6 +6,39 @@
 
     Core sound interface functions and definitions.
 
+****************************************************************************
+
+	In MAME, sound is represented as a graph of sound "streams". Each
+	stream has a fixed number of inputs and outputs, and is responsible
+	for producing sound on demand.
+
+	The graph is driven from the outputs, which are speaker devices.
+	These devices are updated on a regular basis (~50 times per second),
+	and when an update occurs, the graph is walked from the speaker
+	through each input, until all connected streams are up to date.
+
+	Individual streams can also be updated manually. This is important
+	for sound chips and CPU-driven devices, who should force any
+	affected streams to update prior to making changes.
+
+	Sound streams are *not* part of the device execution model. This is
+	very important to understand. If the process of producing the ouput
+	stream affects state that might be consumed by an executing device
+	(e.g., a CPU), then care must be taken to ensure that the stream is
+	updated frequently enough
```

Comment:
From a quick glance over the code changes, this seems to contradict the way the sound-generating devices now use `synchronize` when they want to update their output before a register change takes effect.  To the best of my knowledge, `synchronize` ends the current scheduled device timelice.  How does it interact with the sound streams to get the desired effect?

Reply:
The synchronize changes aren't directly related to this part of the comment. The comment is just a general description of the stream model (which hasn't changed) and how feedback/status from the sound devices is handled. I'll try to explain it better if this is confusing.

In general, for correctness, all writes to sound devices should be synchronized. This is because the base stream_update is run in a timer callback. Synchronizing a write essentially moves it into a timer callback as well, and timer callbacks are guaranteed to run in order, so you never have time slips. If you don't synchronize your writes, then you can introduce situations where the stream update (which should be done prior to applying the write's effects) can pull the stream into the future. For example, if a CPU core executes a few cycles beyond its icount, then machine().time() can be "in the future" relative to a timer callback that will land after the CPU completes its timeslice.

I've deferred bulk moving all sound device writes to synchronized writes for the moment because I think this is a general philosophical question of whether this should be the sound device's responsibility or the driver's responsibility. There is also the general question of whether this matters enough to warrant a bulk change here. My asserts catch it frequently enough to consider.